### ISSUE TITLE
Use Calamus in renku

### ISSUE THREAD

**Issue by jsam** [SwissDataScienceCenter/renku-python#700]
Currently, we have a pretty magical JSON-LD serialization layer with several layers of indirection which is quite a challenge to maintain. To make things more maintainable and more stable I would like to have this as its own library which is properly tested and addresses all the current outstanding tickets (#548, #547).

Some, things to consider might be:
  * Top-level context squashing
  * Flexible way of handling/validating container types
  * Schema migrations
  * Attribute migrations
  * Attribute validations
  * Optionality of attributes
  * (De) Serialization (to and from the given format)

Any draft proposals to address the current situation is more than welcome.

**Comment by Panaetius** [comment]
I think we should restrict ourselves to json-ld 1.1, as it has some nice new feature, such as scoped and nested contexts. If we rely on those features, not using 1.1 will break things.

This means we have to set `@version` in our contexts.

**Comment by Panaetius** [comment]
We should have support for xsd data types in jsonld, so 
```
schema:Something: {
    "@type": "xsd:float",
    "@value": "3.14159"
}
```
gets deserialized correctly as a float value (instead of as a dictionary as in the current implementation)

**Comment by rokroskar** [comment]
This would be awesome!

**Comment by Panaetius** [comment]
General outline of problems we would like to solve (not necessarily specific to a separate library, but also to how it would/should be used in renku-python):
- JSON-LD is slow right now. On import we have to pre-calculate the whole context for all classes, each time, even if those classes aren't really used. Ideally we should only calculate json-ld contexts etc. when `asjsonld` is called (maybe with some caching). This probably means we can't rely on attrs, at least not in the way we currently do.
-  Contexts of children should be propagated to parents at runtime so the json-ld context only contains relevant contexts (instead of always having contexts for all possible children).
- Don't go from commits/activities to entities as we do now. Entities (Dataset etc.) should be the main nodes, annotated with wprov stuff instead of the other way around
- Save json-ld after every renku command, keeping whole graph on disk inside `.renku` somewhere, we don't need to (re-)calculate this stuff every time if it never changes for a commit.
- Save the data as a flat list of nodes that refer to each other by id instead of a nested structure. New entries can just be appended to the end of the file. Possibly split this across multiple files by type, if possible
- Maybe save in expanded form instead of compacted, as it's easier to use and cuts down on processing when running `renku log`. 99% of the time, the json-ld will be read by machines, not humans, so compacted form is not that helpful. We should only show compacted form on demand, e.g. as a flag to `renku log`
- We should support python type annotations and only use manual annotations as a fallback
- Support migrations, possibly automated for simple cases.
- Still support nested contexts and type-scoped contexts
- Use framing (https://json-ld.org/spec/latest/json-ld-framing/ ) in conjunction with the flat list mentioned above.
- Support SHACL
- Generate SHACL? Based on annotations

**Comment by Panaetius** [comment]
For design of the library, I think it'd be nice to split up the single calls to jsonld.

E.g. instead of
```
@jsonld.s(
    type='prov:Activity',
    context={
        'prov': 'http://www.w3.org/ns/prov#',
        'rdfs': 'http://www.w3.org/2000/01/rdf-schema#',
        'schema': 'http://schema.org/'
    },
    cmp=False,
)
```

we could do something like
```
@jsonld_type('prov:Activity')
@jsonld_context([ 'prov': 'http://www.w3.org/ns/prov#',
        'rdfs': 'http://www.w3.org/2000/01/rdf-schema#',
        'schema': 'http://schema.org/'])
```
or for properties it could be `@jsonld_context(...)` for the context and `@reverse()` would turn it into a reverse mapping instead of something like 
```
generated = jsonld.container.list(
        Generation, context={
            '@reverse': 'prov:activity',
        }, kw_only=True
    )
```
(though that's mean you could only decorate @property style properties)

This way we can split up functionality into smaller chunks instead of having the sort of god-methods we have right now that have to handle each and every case.

Then we can also easily ass SHACL validation decorators etc.

**Comment by Panaetius** [comment]
If we use contexts, we need to explicitly unset parent contexts in case of ambiguities, to prevent problems with compaction.

Example:
In the parent context, we have `'_label': 'rdfs:label',` defined. in the child context, it is mapped to `'label': 'rdfs:label',` (notice the missing `_` at the beginning).

Since parent contexts are merged with child contexts, the child context now actually is `{'_label': 'rdfs:label','label': 'rdfs:label'}` and upon compaction, the parent context can be applied (depends on implementation of jsonld processor).

To fix this, we have to set `'_label': None` in the child context so it overrides the parent context.

**TL;DR: We should not rely on contexts....**

**Comment by rokroskar** [comment]
Can we close this? Or at least adapt the epic to include issues from calamus? /cc @Panaetius
